crates/wasi/src/preview2/host/tcp.rs

```diff
@@ -0,0 +1,571 @@
+        }
+    }
+
+    fn start_connect(
+        &mut self,
+        this: tcp::TcpSocket,
+        network: Network,
+        remote_address: IpSocketAddress,
+    ) -> Result<(), network::Error> {
+        let table = self.table();
+        let socket = table.get_tcp_socket(this)?;
+
+        let mut tcp_state = socket.inner.tcp_state.write().unwrap();
+        match &*tcp_state {
+            HostTcpState::Default => {}
+            HostTcpState::Connected => return Err(ErrorCode::AlreadyConnected.into()),
+            _ => return Err(ErrorCode::NotInProgress.into()),
+        }
+
+        let network = table.get_network(network)?;
+        let connecter = network.0.tcp_connecter(remote_address)?;
+
+        // Do a host `connect`. Our socket is non-blocking, so it'll either...
+        match connecter.connect_existing_tcp_listener(socket.tcp_socket()) {
+            // succeed immediately,
+            Ok(()) => {
+                *tcp_state = HostTcpState::ConnectReady(Ok(()));
+                return Ok(());
+            }
+            // continue in progress,
+            Err(err)
+                if err.raw_os_error() == Some(rustix::io::Errno::INPROGRESS.raw_os_error()) => {}
+            // or fail immediately.
+            Err(err) => return Err(err.into()),
+        }
+
+        // The connect is continuing in progres. Set up the join handle.
+
+        let clone = socket.clone_inner();
+
+        #[cfg(unix)]
+        let join = spawn(async move {
+            let result = match clone.tcp_socket.writable().await {
+                Ok(mut writable) => {
+                    writable.retain_ready();
+
+                    // Check whether the connect succeeded.
+                    match sockopt::get_socket_error(&clone.tcp_socket) {
+                        Ok(Ok(())) => Ok(()),
+                        Err(err) | Ok(Err(err)) => Err(err.into()),
+                    }
+                }
+                Err(err) => Err(err),
+            };
+
+            *clone.tcp_state.write().unwrap() = HostTcpState::ConnectReady(result);
+            clone.sender.send(()).unwrap();
+        });
+
+        #[cfg(not(unix))]
+        let join = spawn_blocking(move || {
+            let result = match rustix::event::poll(
+                &mut [rustix::event::PollFd::new(
+                    &clone.tcp_socket,
+                    rustix::event::PollFlags::OUT,
+                )],
+                -1,
+            ) {
+                Ok(_) => {
+                    // Check whether the connect succeeded.
+                    match sockopt::get_socket_error(&clone.tcp_socket) {
+                        Ok(Ok(())) => Ok(()),
+                        Err(err) | Ok(Err(err)) => Err(err.into()),
+                    }
+                }
+                Err(err) => Err(err.into()),
+            };
+
+            *clone.tcp_state.write().unwrap() = HostTcpState::ConnectReady(result);
+            clone.sender.send(()).unwrap();
```

Comment:
Can we consolidate this pattern of setting the new state and then sending () into a single method?